pkg/ignition/enrich.go

```diff
@@ -0,0 +1,154 @@
+)
+
+// Enricher is responsible for enriching the Ignition config with additional data.
+type Enricher struct {
+	BootstrapData []byte
+	Hostname      string
+	InstanceID    string
+	ProviderID    string
+	Network       []cloudinit.NetworkConfigData
+}
+
+// Enrich enriches the Ignition config with additional data.
+func (e *Enricher) Enrich() ([]byte, string, error) {
+	ign, err := e.getEnrichConfig()
+	if err != nil {
+		return nil, "", errors.Wrap(err, "getting enrich config")
+	}
+
+	return buildIgnitionConfig(e.BootstrapData, ign)
+}
+
+func (e *Enricher) getEnrichConfig() (*ignitionTypes.Config, error) {
+	ign := &ignitionTypes.Config{
+		Storage: ignitionTypes.Storage{
+			Files: []ignitionTypes.File{
+				{
+					Node: ignitionTypes.Node{
+						Filesystem: "root",
+						Path:       "/etc/hostname",
+						Overwrite:  ptr.To(true),
+					},
+					FileEmbedded1: ignitionTypes.FileEmbedded1{
+						Mode: ptr.To(0644),
+						Contents: ignitionTypes.FileContents{
+							Source: fmt.Sprintf("data:,%s", e.Hostname),
+						},
+					},
+				},
+				{
+					Node: ignitionTypes.Node{
+						Filesystem: "root",
+						Path:       "/etc/proxmox-env",
+						Overwrite:  ptr.To(true),
+					},
+					FileEmbedded1: ignitionTypes.FileEmbedded1{
+						Mode: ptr.To(420),
+						Contents: ignitionTypes.FileContents{
+							Source: fmt.Sprintf("data:,%s", e.getProxmoxEnvContent()),
+						},
+					},
+				},
+			},
+		},
+		Networkd: ignitionTypes.Networkd{
+			Units: []ignitionTypes.Networkdunit{},
+		},
+		Systemd: ignitionTypes.Systemd{
+			Units: []ignitionTypes.Unit{
+				{
+					Name:   "systemd-resolved.service",
+					Enable: true,
+				},
+			},
+		},
+	}
+
+	// populate networkd units
+	nets, err := RenderNetworkConfigData(cloudinit.BaseCloudInitData{NetworkConfigData: e.Network})
+	if err != nil {
+		return nil, errors.Wrap(err, "rendering networkd units")
+	}
+
+	for name, contents := range nets {
+		ign.Networkd.Units = append(ign.Networkd.Units, ignitionTypes.Networkdunit{
+			Name:     name,
+			Contents: string(contents),
+		})
+	}
+
+	return ign, nil
+}
+
+func (e *Enricher) getProxmoxEnvContent() string {
+	content := fmt.Sprintf("COREOS_CUSTOM_HOSTNAME=%s\nCOREOS_CUSTOM_INSTANCE_ID=%s\nCOREOS_CUSTOM_PROVIDER_ID=%s", e.Hostname, e.InstanceID, e.ProviderID)
+	if len(e.Network) > 0 && e.Network[0].IPAddress != "" {
+		content += fmt.Sprintf("\nCOREOS_CUSTOM_PRIVATE_IPV4=%s", e.Network[0].IPAddress)
+	}
+	if len(e.Network) > 0 && e.Network[0].IPV6Address != "" {
+		content += fmt.Sprintf("\nCOREOS_CUSTOM_PRIVATE_IPV6=%s", e.Network[0].IPV6Address)
+	}
+	return url.PathEscape(content)
+}
+
+func buildIgnitionConfig(bootstrapData []byte, enrichConfig *ignitionTypes.Config) ([]byte, string, error) {
+	// We control bootstrapData config, so treat it as strict.
+	ign, _, err := convertToIgnition(bootstrapData, false)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "converting bootstrap-data to Ignition")
+	}
+
+	var clcWarnings string
```

Comment:
This looks like development leftovers. Since it's returned by value, I don't see the point of this at all.

Reply:
the clcWarnings are returned 
so when something is wrong in the ignition it reports the warnings and errors.
it's good for logging

Reply:
clcWarnings aren't initialized here, and they're passed back uninitialised. that doesn't really make sense, unless there's something I'm missing.

Reply:
you're right

Reply:
removed